crates/kubelet/src/node.rs

```diff
@@ -166,6 +170,140 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
+pub async fn drain_node(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    cordon_node(client, node_name).await?;
+    evict_pods(client, node_name).await?;
+    Ok(())
+}
+
+/// Fetches list of pods on this node and deletes them.
+pub async fn evict_pods(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    let pod_client: Api<KubePod> = Api::all(client.clone());
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    let lp = ListParams::default().fields(&format!("spec.nodeName={}", node_name));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    let mut stream = pod_client.watch(&lp, "0").await?.boxed();
+
+    info!("Evicting {} pods.", pods.len());
+
+    for pod in pods {
+        let pod = Pod::new(pod);
+        if pod.is_daemonset() {
+            info!("Skipping eviction of DaemonSet '{}'", pod.name());
+            continue;
+        } else if pod.is_static() {
+            let mut container_statuses = HashMap::new();
+            for container in pod.containers() {
+                container_statuses.insert(
+                    container.name.clone(),
+                    ContainerStatus::Terminated {
+                        timestamp: Utc::now(),
+                        message: "Evicted on node shutdown.".to_string(),
+                        failed: false,
+                    },
+                );
+            }
+            let status = Status {
+                message: Some("Evicted on node shutdown.".to_string()),
+                container_statuses,
+            };
+            pod.patch_status(client.clone(), status).await;
+            info!("Marked static pod as terminated.");
+            continue;
+        } else {
+            match evict_pod(&client, pod.name(), pod.namespace(), &mut stream).await {
+                Ok(_) => (),
+                Err(e) => {
+                    // Absorb the error and attempt to delete other pods with best effort.
+                    error!("Error evicting pod: {:?}", e)
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+type PodStream = std::pin::Pin<
+    Box<
+        dyn futures::Stream<Item = Result<kube::api::WatchEvent<KubePod>, kube::error::Error>>
+            + Send,
+    >,
+>;
+
+async fn evict_pod(
+    client: &kube::Client,
+    name: &str,
+    namespace: &str,
+    stream: &mut PodStream,
+) -> anyhow::Result<()> {
+    let ns_client: Api<KubePod> = Api::namespaced(client.clone(), namespace);
+    info!("Evicting namespace '{}' pod '{}'", namespace, name);
+    let params = Default::default();
+    let response = ns_client.delete(name, &params).await?;
+
+    if response.is_left() {
+        // TODO Timeout?
```

Comment:
If you want to, this can be done with the `retry!` macro in this file. Not going to block on it though